compiler/src/dotty/tools/dotc/interactive/Completion.scala

```diff
@@ -185,32 +223,35 @@ object Completion {
      * If `sym` exists, no symbol with the same name is already included, and it satisfies the
      * inclusion filter, then add it to the completions.
      */
-    private def add(sym: Symbol)(implicit ctx: Context) =
-      if (sym.exists && !completions.lookup(sym.name).exists && include(sym)) {
-        completions.enter(sym)
+    private def add(sym: Symbol, nameInScope: Name)(implicit ctx: Context) =
+      if (sym.exists && !completions.lookup(nameInScope).exists && include(sym, nameInScope)) {
+        completions.enter(sym, nameInScope)
       }
 
     /** Lookup members `name` from `site`, and try to add them to the completion list. */
-    private def addMember(site: Type, name: Name)(implicit ctx: Context) =
-      if (!completions.lookup(name).exists)
-        for (alt <- site.member(name).alternatives) add(alt.symbol)
+    private def addMember(site: Type, name: Name, nameInScope: Name)(implicit ctx: Context) =
+      if (!completions.lookup(nameInScope).exists) {
+        for (alt <- site.member(name).alternatives) add(alt.symbol, nameInScope)
+      }
 
     /** Include in completion sets only symbols that
      *   1. start with given name prefix, and
      *   2. do not contain '$' except in prefix where it is explicitly written by user, and
      *   3. are not a primary constructor,
-     *   4. are the module class in case of packages,
-     *   5. are mutable accessors, to exclude setters for `var`,
-     *   6. have same term/type kind as name prefix given so far
+     *   4. have an existing source symbol,
+     *   5. are the module class in case of packages,
+     *   6. are mutable accessors, to exclude setters for `var`,
+     *   7. have same term/type kind as name prefix given so far
      *
      *  The reason for (2) is that we do not want to present compiler-synthesized identifiers
      *  as completion results. However, if a user explicitly writes all '$' characters in an
      *  identifier, we should complete the rest.
      */
-    private def include(sym: Symbol)(implicit ctx: Context): Boolean =
-      sym.name.startsWith(prefix) &&
-      !sym.name.toString.drop(prefix.length).contains('$') &&
+    private def include(sym: Symbol, nameInScope: Name)(implicit ctx: Context): Boolean =
+      nameInScope.startsWith(prefix) &&
+      !nameInScope.toString.drop(prefix.length).contains('$') &&
```

Comment:
Still not sure I understand why this is needed, can we remove it and see what fails ? :)

Reply:
See this comment:
```
    /** Include in completion sets only symbols that
     *   1. start with given name prefix, and
     *   2. do not contain '$' except in prefix where it is explicitly written by user, and
     *   (...)
     *
     *  The reason for (2) is that we do not want to present compiler-synthesized identifiers
     *  as completion results. However, if a user explicitly writes all '$' characters in an
     *  identifier, we should complete the rest.
     */
```

We would have to remove the `object Foo$`s, `def foo$default$1`, etc.

Reply:
This was dead code since #3960, let me see how to plug it back in.